sections/sections-app/src/java/org/sakaiproject/tool/section/jsf/backingbean/RosterBean.java

```diff
@@ -121,50 +125,71 @@ public void init() {
-		for(Iterator<EnrollmentRecord> studentIter = siteStudents.iterator(); studentIter.hasNext();) {
-			EnrollmentRecord enrollment = studentIter.next();
-
+		unpagedEnrollments = new ArrayList<>();
+		siteStudents.forEach(enrollment -> {
 			// Build a map of categories to sections in which the student is enrolled
-			Map<String, CourseSection> map = new HashMap<String, CourseSection>();
-			for(Iterator catIter = categories.iterator(); catIter.hasNext();) {
-				String cat = (String)catIter.next();
-				CourseSection section = sectionEnrollments.getSection(enrollment.getUser().getUserUid(), cat);
-                map.put(cat, section);
-			}
+			Map<String, CourseSection> studentSectionsMap = new HashMap<>();
 
-			// Check to see whether this enrollment should be filtered out
-			boolean includeStudent = false;
-			if(StringUtils.trimToNull(getFilter()) == null) {
-				includeStudent = true;
-			} else {
-				for(Iterator<Entry<String, CourseSection>> entryIter = map.entrySet().iterator(); entryIter.hasNext();) {
-					Entry<String, CourseSection> entry = entryIter.next();
-					CourseSection section = entry.getValue();
-					// Some map entries won't have a section at all, since the student isn't in any section of that category
-					if(section == null) {
-						continue;
-					}
-					if("MY".equals(getFilter()) && assignedSections.contains(section)) {
-						includeStudent = true;
-						break;
-					} else if(section.getCategory().equals(getFilter()) && assignedSections.contains(section)) {
-						includeStudent = true;
-						break;
+			// Iterate site sections to see if the student is a member of and the filter matches the selected section
+			boolean includeStudent = categories.stream().filter(category -> {
+				CourseSection section = sectionEnrollments.getSection(enrollment.getUser().getUserUid(), category);
+				boolean isValidSection = false;
+
+				if (section != null) {
+					if(("MY".equals(getFilter()) || section.getCategory().equals(getFilter()))
+							&& assignedSections.contains(section)) {
+						isValidSection = true;
+
+					} else if (StringUtils.isNotBlank(section.getEid()) && section.getEid().equals(getFilter())) {
+						isValidSection = true;
+
+					} else if (StringUtils.isNotBlank(section.getTitle()) && section.getTitle().equals(getFilter())) {
+						isValidSection = true;
 					}
+
+					studentSectionsMap.put(category, section);
 				}
+
+				return isValidSection;
+			}).count() > 0; // If there are valid sections for the current filter, include the student in the view
+
+			// If there is no filter, the section is valid and the student shouldn't be excluded
+			if (StringUtils.isBlank(getFilter())) {
+				includeStudent = true;
+			}
+
+			// If the filter is set to "no sections" filter and the student sections map is empty, student should be included
+			if (NO_SECTIONS_FILTER_KEY.equals(getFilter()) && studentSectionsMap.isEmpty()) {
+				includeStudent = true;
```

Comment:
I would put both these checks before the categories lambda, and only run that lambda if includeStudent is false at that point. Also, these two blocks could also probably be combined into one.

includeStudent = StringUtils.isBlank(getFilter()) || (NO_SECTIONS_FILTER_KEY.equals(getFilter()) && studentSectionsMap.isEmpty());

Reply:
Hi @adrianfish , thanks for your suggestions, but this one and the `anyMatch();` one, will do this not to work as expected. I also thought about using anymatch(), but `studentSectionsMap` is fulfilled in the stream, so if we broke the stream using `anyMatch();`, the map could be incomplete. Also, checking this map should be done after the stream, because this variable is fulfilled in the stream.

Your other suggestions seems good to me and I've implemented them, thanks!

Reply:
Understood.